File: mac/WisprMac/HistoryManager.swift
import Foundation

struct HistoryItem: Identifiable, Codable {
    let id: UUID
    let text: String
    let timestamp: Date
}

class HistoryManager {
    static let shared = HistoryManager()
    
    private let key = "dictation_history"
    
    func saveItem(text: String) {
        var items = getHistory()
        let newItem = HistoryItem(id: UUID(), text: text, timestamp: Date())
        items.insert(newItem, at: 0) // Newest first
        
        // Keep only items from the last 24 hours (86400 seconds)
        let cutoff = Date().addingTimeInterval(-86400)
        items = items.filter { $0.timestamp > cutoff }
        
        if let encoded = try? JSONEncoder().encode(items) {
            UserDefaults.standard.set(encoded, forKey: key)
        }
        
        // Notify SettingsView to refresh
        NotificationCenter.default.post(name: NSNotification.Name("WisprHistoryChanged"), object: nil)
    }
    
    func getHistory() -> [HistoryItem] {
        guard let data = UserDefaults.standard.data(forKey: key) else { return [] }
        if let decoded = try? JSONDecoder().decode([HistoryItem].self, from: data) {
            let cutoff = Date().addingTimeInterval(-86400)
            return decoded.filter { $0.timestamp > cutoff }
        }
        return []
    }
    
    func clearHistory() {
        UserDefaults.standard.removeObject(forKey: key)
        NotificationCenter.default.post(name: NSNotification.Name("WisprHistoryChanged"), object: nil)
    }
}

File: android/app/src/main/java/com/wispr/HistoryManager.kt
package com.wispr

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

data class HistoryItem(
    val id: String,
    val text: String,
    val timestamp: Long
)

class HistoryManager(context: Context) {
    private val prefs = context.getSharedPreferences("wispr_prefs", Context.MODE_PRIVATE)
    private val key = "dictation_history"

    fun saveItem(text: String) {
        val items = getHistory().toMutableList()
        val newItem = HistoryItem(
            id = UUID.randomUUID().toString(),
            text = text,
            timestamp = System.currentTimeMillis()
        )
        items.add(0, newItem) // Newest first

        // Keep only items from the last 24 hours (86400 * 1000 ms)
        val cutoff = System.currentTimeMillis() - 86400 * 1000
        val filtered = items.filter { it.timestamp > cutoff }

        saveList(filtered)
    }

    fun getHistory(): List<HistoryItem> {
        val jsonString = prefs.getString(key, null) ?: return emptyList()
        try {
            val jsonArray = JSONArray(jsonString)
            val list = mutableListOf<HistoryItem>()
            for (i in 0 until jsonArray.length()) {
                val obj = jsonArray.getJSONObject(i)
                list.add(
                    HistoryItem(
                        id = obj.getString("id"),
                        text = obj.getString("text"),
                        timestamp = obj.getLong("timestamp")
                    )
                )
            }
            // Filter out items older than 24 hours
            val cutoff = System.currentTimeMillis() - 86400 * 1000
            return list.filter { it.timestamp > cutoff }
        } catch (e: Exception) {
            e.printStackTrace()
            return emptyList()
        }
    }

    fun clearHistory() {
        prefs.edit().remove(key).apply()
    }

    private fun saveList(list: List<HistoryItem>) {
        val jsonArray = JSONArray()
        for (item in list) {
            val obj = JSONObject().apply {
                put("id", item.id)
                put("text", item.text)
                put("timestamp", item.timestamp)
            }
            jsonArray.put(obj)
        }
        prefs.edit().putString(key, jsonArray.toString()).apply()
    }
}
